text/3759-cargo-supported-targets.md

```diff
@@ -0,0 +1,517 @@
+the door for more advanced control over dependencies.
+For example, transitive dependencies declared under a `[target.**.dependencies]` table could be
+excluded from `Cargo.lock` if the dependent's `supported-targets` is mutually exclusive with
+the target preconditions under which the dependencies are included.
+This is especially relevant to areas such as WebAssembly and embedded programming,
+where one usually supports only a few specific targets. Currently, auditing and vendoring
+is tedious because dependencies under `[target.**.dependencies]` tables always make their way
+in the dependency tree, even though they may not actually be used.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `supported-targets` field can be added to `Cargo.toml` under the `[package]` table.
+
+This field consists of an array of strings, where each string is an explicit target-triple or a `cfg` specification
+(as for the `[target.'cfg(**)']` table). The supported `cfg` syntax is the same as the one for
+[platform-specific dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies)
+(i.e., `cfg(test)`, `cfg(debug_assertions)`, and `cfg(proc_macro)` are not supported).
+If a selected target satisfies any entry of the `supported-targets` list, then the package can be built for that target.
+
+__For example:__
+```toml
+[package]
+name = "hello_cargo"
+version = "0.1.0"
+edition = "2021"
+supported-targets = [
+    "wasm32-unknown-unknown",
+    'cfg(target_os = "linux")',
+    'cfg(target_os = "macos")'
+]
+```
+Here, only targets satisfying: the `wasm32-unknown-unknown` target, __or__ the `linux` OS, __or__ the `macos` OS,
+are allowed to build the package. If no `supported-targets` was specified, then any target would be allowed.
+
+User experience is enhanced by raising an error that fails compilation when the supported targets
+of a package are not satisfied by the selected target. A package's `supported-targets` must be a subset
+of its dependencies' `supported-targets`, otherwise the build also fails.
+
+When `supported-targets` is not specified, any target is accepted, so all dependencies must support
+all targets.
+
+This feature should be used when a package clearly does not support all targets. For example: `io-uring`
+requires `cfg(target_os = "linux")`, `gloo` requires `cfg(target_family = "wasm")`, and
+`riscv` requires `cfg(target_arch = "riscv32")` or `cfg(target_arch = "riscv64")`.
+
+This feature should also be used to increase cargo's knowledge of a package. For example,
+when working in a workspace where some packages compile with `#[no_std]` and `target_os = "none"`, 
+and some others are tools that require a desktop OS, using `supported-targets` makes
+`cargo <command>` ignore packages which have `supported-targets` that are not satisfied
+by the selected target.
+
```

Comment:
```suggestion

Cargo's documentation should give clear guidance for when to use this field, and should not suggest using it by default. In particular, we should steer users to use this when they have good reason to believe the crate will not compile or work as expected (e.g. because it uses target-specific APIs), and not use it merely for "I haven't personally tested this on other targets".

```